File: 112.Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    if (!root) {
        return false;
    }
    if (!root.left && !root.right) {
        return root.val === targetSum;
    }
    return hasPathSum(root.left, targetSum - root.val) || hasPathSum(root.right, targetSum - root.val);
};

var hasPathSum = function(root, targetSum) {
    if (!root) {
        return false;
    }
    const stack = [[root, targetSum]];
    while(stack.length) {
        const [node, num] = stack.pop();
        const diff = num - node.val;
        if (!node.left && !node.right && diff === 0) {
            return true;   
        }
        if (node.left) {
            stack.push([node.left, diff]);   
        }
        if (node.right) {
            stack.push([node.right, diff]);   
        }
    }
    return false;
};

File: Golang/Tree/112.Path-sum.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSumRecursive(root *TreeNode, targetSum int) bool {
	var recursive func(node *TreeNode, value int) bool
	recursive = func(node *TreeNode, value int) bool {
		if node == nil {
			return false
		}
		if node.Val == value && node.Left == nil && node.Right == nil {
			return true
		}
		return recursive(node.Left, value-node.Val) || recursive(node.Right, value-node.Val)
	}
	return recursive(root, targetSum)
}

func hasPathSumBFS(root *TreeNode, targetSum int) bool {
	if root != nil {
		type ComplexNode struct {
			Node   *TreeNode
			Target int
		}
		stack := []ComplexNode{{
			Node:   root,
			Target: targetSum,
		}}
		for len(stack) > 0 {
			children := []ComplexNode{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				if node.Node.Val == node.Target && node.Node.Left == nil && node.Node.Right == nil {
					return true
				}
				if node.Node.Left != nil {
					children = append(children, ComplexNode{
						Node:   node.Node.Left,
						Target: node.Target - node.Node.Val,
					})
				}
				if node.Node.Right != nil {
					children = append(children, ComplexNode{
						Node:   node.Node.Right,
						Target: node.Target - node.Node.Val,
					})
				}
			}
			stack = children
		}
	}
	return false
}
